### PR TITLE
Add integration tests for Queue controller

### PR DESCRIPTION
**Pull request by ArangoGutierrez**
Signed-off-by: Carlos Eduardo Arango Gutierrez <carangog@redhat.com>

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
As requested in #88

### CODE REVIEW

File: test/integration/controller/core/queue_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+var _ = ginkgo.Describe("Queue controller", func() {
+	var (
+		ns                *corev1.Namespace
+		Queue             *kueue.Queue
+		clusterQueue      *kueue.ClusterQueue
+		emptyQueueStatus  kueue.QueueStatus
+		loadedQueueStatus kueue.QueueStatus
+	)
+
+	loadedQueueStatus = kueue.QueueStatus{PendingWorkloads: 3}
```
Comment (ahg-g) [inline test/integration/controller/core/queue_controller_test.go:42]: doesn't look necessary to define those here, just inline them.

File: test/integration/controller/core/queue_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+		ginkgo.By("Creating workloads")
+		for _, w := range workloads {
+			gomega.Expect(k8sClient.Create(ctx, w)).To(gomega.Succeed())
+		}
+		gomega.Eventually(func() kueue.QueueStatus {
+			var updatedQueue kueue.Queue
+			gomega.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(Queue), &updatedQueue)).To(gomega.Succeed())
+			return updatedQueue.Status
+		}, framework.Timeout, framework.Interval).Should(testing.Equal(loadedQueueStatus))
```
Comment (ahg-g) [inline test/integration/controller/core/queue_controller_test.go:89]:
```suggestion
		}, framework.Timeout, framework.Interval).Should(testing.Equal(kueue.QueueStatus{PendingWorkloads: 3}))
```

File: test/integration/controller/core/queue_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+		ginkgo.By("Finishing workloads")
+		for _, w := range workloads {
+			w.Status.Conditions = append(w.Status.Conditions, kueue.QueuedWorkloadCondition{
```
Comment (ahg-g) [inline test/integration/controller/core/queue_controller_test.go:93]: we just need to "admit" the workloads, not finish them to have the PendingWorkloads status go to zero, right? can we do this test by setting the "Admission" field on the workload instead?

File: test/integration/controller/core/queue_controller_test.go (hunk 1 above)
Comment (ArangoGutierrez) [reply test/integration/controller/core/queue_controller_test.go:42]: sure why not

File: test/integration/controller/core/queue_controller_test.go (hunk 3 above)
Comment (ArangoGutierrez) [reply test/integration/controller/core/queue_controller_test.go:93]: basically "any" condition that is not "Assigned" will set the "Pending" to 0 , so really doesn't matter here

**Comment by ArangoGutierrez** [comment]
> @ArangoGutierrez: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-kueue-verify-main	[0afb5f5](https://github.com/kubernetes-sigs/kueue/commit/0afb5f56781e33366824a358456062d5171e848a)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kueue/98/pull-kueue-verify-main/1501003006076784640)	true	`/test pull-kueue-verify-main`
> pull-kueue-test-unit-main	[0afb5f5](https://github.com/kubernetes-sigs/kueue/commit/0afb5f56781e33366824a358456062d5171e848a)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kueue/98/pull-kueue-test-unit-main/1501003005992898560)	true	`/test pull-kueue-test-unit-main`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes-sigs&repo=kueue&pr=98). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3AArangoGutierrez). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes-sigs/kueue/issues?q=is:issue+is:open) when you hit one in your PR.

Why can commitSuggestion ever work ;P ... fixing and pushing... 1 min

File: test/integration/controller/core/queue_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+				Type:   kueue.QueuedWorkloadFinished,
+				Status: corev1.ConditionTrue,
+			})
+			gomega.Expect(k8sClient.Status().Update(ctx, w)).To(gomega.Succeed())
+		}
+		gomega.Eventually(func() kueue.QueueStatus {
+			var updatedQueue kueue.Queue
+			gomega.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(Queue), &updatedQueue)).To(gomega.Succeed())
+			return updatedQueue.Status
+		}, framework.Timeout, framework.Interval).Should(testing.Equal(emptyQueueStatus))
```
Comment (ahg-g) [inline test/integration/controller/core/queue_controller_test.go:100]:
```suggestion
		}, framework.Timeout, framework.Interval).Should(testing.Equal(kueue.QueueStatus{}))
```

and we can remove the variable

Comment (ArangoGutierrez) [reply test/integration/controller/core/queue_controller_test.go]: sure
done

File: test/integration/controller/core/queue_controller_test.go (hunk 3 above)
Comment (ahg-g) [reply test/integration/controller/core/queue_controller_test.go:89]: hmm, it seems that we are missing to set the QW condition to `QueuedWorkloadAdmitted` when the workload gets admitted, so currently the status will continue to count the workload as pending while it is running, but this is not related to this integration test...

**Comment by ahg-g** [comment]
looks good, please squash.

File: test/integration/controller/core/queue_controller_test.go (hunk 3 above)
Comment (ArangoGutierrez) [reply test/integration/controller/core/queue_controller_test.go:89]: yup, the k.queue.status.pending just reflects QW that have not been admitted yet , what ever happens after that is not yet reported into the k.queue.status

Comment (ArangoGutierrez) [reply test/integration/controller/core/queue_controller_test.go:89]: that's actually the WIP of #88 , adding more things to report there

**Comment by ArangoGutierrez** [comment]
> looks good, please squash.

Done :)

File: test/integration/controller/core/queue_controller_test.go (hunk 3 above)
Comment (ahg-g) [reply test/integration/controller/core/queue_controller_test.go:89]: actually the admission status is not based on the condition, it is based on the "admission" field being set, I think we need a test for that, basically have some workloads finish and some only getting the admission field set and that should impact the PendingWorkloads status

Comment (ahg-g) [reply test/integration/controller/core/queue_controller_test.go:89]: see https://github.com/kubernetes-sigs/kueue/blob/28351208f24e5a17337bde997950209affc11f01/pkg/controller/core/queuedworkload_controller.go#L189

Comment (ahg-g) [reply test/integration/controller/core/queue_controller_test.go:89]: > that's actually the WIP of #88 , adding more things to report there

we need to set the condition in the scheduler in this go routine: https://github.com/kubernetes-sigs/kueue/blob/28351208f24e5a17337bde997950209affc11f01/pkg/scheduler/scheduler.go#L223

Comment (ArangoGutierrez) [reply test/integration/controller/core/queue_controller_test.go:89]: I can add that on #88 for sure

File: api/v1alpha1/queue_types.go [hunk 5]
```diff
@@ -31,8 +31,8 @@ type ClusterQueueReference string
 
 // QueueStatus defines the observed state of Queue
 type QueueStatus struct {
-	// PendingWorkloads is the number of workloads currently assigned to this
-	// queue not yet assigned to a ClusterQueue.
+	// PendingWorkloads is the number of workloads currently admitted to this
+	// queue not yet admitted to a ClusterQueue.
```
Comment (alculquicondor) [inline api/v1alpha1/queue_types.go:35]: in this queue not yet admitted to a ClusterQueue?

File: test/integration/controller/core/queue_controller_test.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+				Flavor(testing.MakeFlavor(flavorModelA, "5").Ceiling("10").Obj()).
+				Flavor(testing.MakeFlavor(flavorModelB, "5").Ceiling("10").Obj()).Obj()).Obj()
+		Queue = testing.MakeQueue("queue", ns.Name).
+			Capacity("cluster-queue").Obj()
+		gomega.Expect(k8sClient.Create(ctx, Queue)).To(gomega.Succeed())
+	})
+
+	ginkgo.AfterEach(func() {
+		gomega.Expect(framework.DeleteQueue(ctx, k8sClient, Queue)).To(gomega.Succeed())
+		gomega.Expect(framework.DeleteClusterQueue(ctx, k8sClient, clusterQueue)).To(gomega.Succeed())
```
Comment (alculquicondor) [inline test/integration/controller/core/queue_controller_test.go:61]: DeleteNamespace too

File: test/integration/controller/core/queue_controller_test.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	"sigs.k8s.io/kueue/pkg/util/testing"
+	"sigs.k8s.io/kueue/test/integration/framework"
+)
+
+// +kubebuilder:docs-gen:collapse=Imports
+
+var _ = ginkgo.Describe("Queue controller", func() {
+	var (
+		ns           *corev1.Namespace
+		Queue        *kueue.Queue
```
Comment (alculquicondor) [inline test/integration/controller/core/queue_controller_test.go:36]: `queue`?

File: test/integration/controller/core/queue_controller_test.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+		for _, w := range workloads {
+			gomega.Expect(k8sClient.Create(ctx, w)).To(gomega.Succeed())
+		}
+		gomega.Eventually(func() kueue.QueueStatus {
+			var updatedQueue kueue.Queue
+			gomega.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(Queue), &updatedQueue)).To(gomega.Succeed())
+			return updatedQueue.Status
+		}, framework.Timeout, framework.Interval).Should(testing.Equal(kueue.QueueStatus{PendingWorkloads: 3}))
+
+		ginkgo.By("Finishing workloads")
```
Comment (alculquicondor) [inline test/integration/controller/core/queue_controller_test.go:87]: What about admitting?

File: api/v1alpha1/queue_types.go (hunk 5 above)
Comment (ArangoGutierrez) [reply api/v1alpha1/queue_types.go:35]: Will address in a followup PR

File: test/integration/controller/core/queue_controller_test.go (hunk 6 above)
Comment (ArangoGutierrez) [reply test/integration/controller/core/queue_controller_test.go:61]: Will address in a followup PR

File: test/integration/controller/core/queue_controller_test.go (hunk 7 above)
Comment (ArangoGutierrez) [reply test/integration/controller/core/queue_controller_test.go:36]: Will address in a followup PR

File: test/integration/controller/core/queue_controller_test.go (hunk 8 above)
Comment (ArangoGutierrez) [reply test/integration/controller/core/queue_controller_test.go:87]: Will address in a followup PR

**Comment by ArangoGutierrez** [comment]
https://github.com/kubernetes-sigs/kueue/issues/68

**Pull request merged**